test/e2e_node/restart_test.go

```diff
@@ -265,5 +269,103 @@ var _ = SIGDescribe("Restart [Serial] [Slow] [Disruptive]", func() {
 				}
 			}
 		})
+		// Regression test for https://issues.k8s.io/116925
+		ginkgo.It("should delete pods which are marked as terminal and have a deletion timestamp set after restart", func(ctx context.Context) {
+			podName := "terminal-restart-pod" + string(uuid.NewUUID())
+			gracePeriod := int64(30)
+			podSpec := e2epod.MustMixinRestrictedPodSecurity(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: podName,
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &gracePeriod,
+					RestartPolicy:                 v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						{
+							Name:    podName,
+							Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+							Command: []string{"sh", "-c"},
+							Args: []string{`
+							_term() {
+								echo "Caught SIGTERM signal!"
+								sleep 5
+								exit 0
+							}
+							trap _term SIGTERM
+							while true; do sleep 5; done
```

Comment:
That's a good point and not necessary intentional :). It does not really matter for this test case, because if SIGTERM is sent, as soon as the sleep is finished (every 5sec), the term handler will run. 

In either case, it's unnecessary to wait for the sleep to be over, so I rewrote this to trap `SIGTERM` instead and `wait` on child sleep pid which results in the term trap being run immediately.